File: syntecnia/core/tokens.py
"""
Syntecnia Token Definitions.

Tokens are the atomic units of the language. Syntecnia's tokens reflect
its design philosophy: readable, flat, intention-based.

The language uses natural-flow keywords instead of traditional programming
constructs:
    - 'when' instead of 'if'
    - 'otherwise' instead of 'else'
    - 'each' instead of 'for'
    - 'task' instead of 'function'
    - 'require' for capability declarations
    - 'approve' for human intervention points
    - 'agent' for agent definitions
    - 'share' for blackboard/shared state
    - 'reason' for LLM-powered reasoning blocks
    - 'invariant' for runtime guarantees
    - 'intent' for declaring the purpose of an operation
"""

from enum import Enum, auto
from dataclasses import dataclass
from typing import Any, Optional


class TokenType(Enum):
    # -- Literals --
    NUMBER = auto()          # 42, 3.14
    TEXT = auto()             # "hello", 'world'
    BOOL_TRUE = auto()       # true
    BOOL_FALSE = auto()      # false
    NOTHING = auto()         # nothing (null equivalent)
    IDENTIFIER = auto()      # variable/function names

    # -- Flow keywords (replace traditional if/else/for) --
    WHEN = auto()            # conditional: when condition
    OTHERWISE = auto()       # else branch: otherwise
    EACH = auto()            # iteration: each item in list
    WHILE = auto()           # loop: while condition
    MATCH = auto()           # pattern matching: match value
    IS = auto()              # match arm: is pattern
    THEN = auto()            # consequence: then action
    AND_THEN = auto()        # sequential step: and then
    STOP = auto()            # early exit: stop
    REPEAT = auto()          # loop restart: repeat
    IN = auto()              # membership: item in collection
    WITH = auto()            # context/binding: with x as y

    # -- Definition keywords --
    TASK = auto()            # function: task name(params)
    GIVE = auto()            # return: give value
    LET = auto()             # binding: let x be value
    BE = auto()              # assignment: let x be 5
    SET = auto()             # mutation: set x to 5
    TO = auto()              # target: set x to value
    AS = auto()              # aliasing: with x as y
    OF = auto()              # property access: name of person
    TYPE = auto()            # type definition: type Name

    # -- Agent keywords --
    AGENT = auto()           # agent definition
    SPAWN = auto()           # create sub-agent
    SHARE = auto()           # publish to blackboard
    OBSERVE = auto()         # read from blackboard
    STATE = auto()           # agent state declaration
    SIGNAL = auto()          # inter-agent signal
    WAIT_FOR = auto()        # wait for signal/condition

    # -- Capability & Security keywords --
    REQUIRE = auto()         # capability declaration
    ALLOW = auto()           # grant permission
    DENY = auto()            # revoke permission
    SANDBOX = auto()         # sandboxed execution block
    VERIFY = auto()          # verify invariant

    # -- Human interaction keywords --
    APPROVE = auto()         # human approval point
    SHOW = auto()            # display/preview to human
    ASK = auto()             # ask human for input
    CONFIRM = auto()         # confirmation checkpoint

    # -- Reasoning keywords (LLM integration) --
    REASON = auto()          # LLM reasoning block
    INTENT = auto()          # declare operation intent
    INVARIANT = auto()       # runtime guarantee
    DECIDE = auto()          # LLM-powered decision
    ANALYZE = auto()         # LLM analysis
    GENERATE = auto()        # LLM content generation

    # -- Error handling keywords --
    TRY = auto()             # try block
    RECOVER = auto()         # catch/recover block

    # -- HTTP server SOFT keywords --
    # These are NOT in KEYWORDS: the lexer emits them as IDENTIFIER and the
    # parser recognizes them only at the start of their construction (serve on
    # N, route "...", requires auth, expect body {...}) via fixed lookahead.
    # The enum members are kept as documentation of the reserved-by-context set.
    SERVE = auto()           # serve on PORT — start an HTTP server
    ON = auto()              # serve on PORT
    ROUTE = auto()           # route "GET /path" — define a route
    AUTH = auto()            # auth with <task> / requires auth
    REQUIRES = auto()        # route ... requires auth
    EXPECT = auto()          # expect body {field: type} — input validation

    # -- Observability keywords --
    TRACE = auto()           # tracing block
    LOG = auto()             # log emission
    MEASURE = auto()         # performance measurement
    CHECKPOINT = auto()      # state snapshot

    # -- Operators --
    PLUS = auto()            # +
    MINUS = auto()           # -
    STAR = auto()            # *
    SLASH = auto()           # /
    PERCENT = auto()         # %
    POWER = auto()           # **

    EQUAL = auto()           # ==
    NOT_EQUAL = auto()       # !=
    LESS = auto()            # <
    GREATER = auto()         # >
    LESS_EQUAL = auto()      # <=
    GREATER_EQUAL = auto()   # >=

    AND = auto()             # and
    OR = auto()              # or
    NOT = auto()             # not

    ASSIGN = auto()          # =
    ARROW = auto()           # ->
    FAT_ARROW = auto()       # =>
    PIPE = auto()            # |>

    # -- Delimiters --
    LPAREN = auto()          # (
    RPAREN = auto()          # )
    LBRACKET = auto()        # [
    RBRACKET = auto()        # ]
    LBRACE = auto()          # {  (used only for maps)
    RBRACE = auto()          # }
    COMMA = auto()           # ,
    DOT = auto()             # .
    COLON = auto()           # :
    NEWLINE = auto()         # \n (significant in Syntecnia)
    INDENT = auto()          # indentation increase
    DEDENT = auto()          # indentation decrease

    # -- Special --
    COMMENT = auto()         # -- comment
    EOF = auto()             # end of file
    ERROR = auto()           # lexer error token


# Keywords mapping
KEYWORDS = {
    # Flow
    "when": TokenType.WHEN,
    "otherwise": TokenType.OTHERWISE,
    "each": TokenType.EACH,
    "while": TokenType.WHILE,
    "match": TokenType.MATCH,
    "is": TokenType.IS,
    "then": TokenType.THEN,
    "and_then": TokenType.AND_THEN,
    "stop": TokenType.STOP,
    "repeat": TokenType.REPEAT,
    "in": TokenType.IN,
    "with": TokenType.WITH,

    # Definitions
    "task": TokenType.TASK,
    "give": TokenType.GIVE,
    "let": TokenType.LET,
    "be": TokenType.BE,
    "set": TokenType.SET,
    "to": TokenType.TO,
    "as": TokenType.AS,
    "of": TokenType.OF,
    "type": TokenType.TYPE,

    # Agent
    "agent": TokenType.AGENT,
    "spawn": TokenType.SPAWN,
    "share": TokenType.SHARE,
    "observe": TokenType.OBSERVE,
    "state": TokenType.STATE,
    "signal": TokenType.SIGNAL,
    "wait_for": TokenType.WAIT_FOR,

    # Capabilities
    "require": TokenType.REQUIRE,
    "allow": TokenType.ALLOW,
    "deny": TokenType.DENY,
    "sandbox": TokenType.SANDBOX,
    "verify": TokenType.VERIFY,

    # Human
    "approve": TokenType.APPROVE,
    "show": TokenType.SHOW,
    "ask": TokenType.ASK,
    "confirm": TokenType.CONFIRM,

    # Reasoning
    "reason": TokenType.REASON,
    "intent": TokenType.INTENT,
    "invariant": TokenType.INVARIANT,
    "decide": TokenType.DECIDE,
    "analyze": TokenType.ANALYZE,
    "generate": TokenType.GENERATE,

    # Observability
    "trace": TokenType.TRACE,
    "log": TokenType.LOG,
    "measure": TokenType.MEASURE,
    "checkpoint": TokenType.CHECKPOINT,

    # Error handling
    "try": TokenType.TRY,
    "recover": TokenType.RECOVER,

    # NOTE: the HTTP-server words (serve, on, route, auth, requires, expect)
    # are intentionally NOT reserved. They are *soft keywords*: the parser
    # recognizes them only at the start of their construction (serve on N,
    # route "...", requires auth, expect body {...}) via fixed lookahead.
    # Everywhere else they are ordinary identifiers, so `let route be "/x"`
    # and `task auth(x)` are valid.

    # Literals
    "true": TokenType.BOOL_TRUE,
    "false": TokenType.BOOL_FALSE,
    "nothing": TokenType.NOTHING,

    # Logical operators
    "and": TokenType.AND,
    "or": TokenType.OR,
    "not": TokenType.NOT,
}


@dataclass(frozen=True)
class SourceLocation:
    """Exact position in source code for error reporting and tracing."""
    file: str
    line: int
    column: int
    offset: int  # byte offset in source

    def __str__(self) -> str:
        return f"{self.file}:{self.line}:{self.column}"


@dataclass(frozen=True)
class Token:
    """
    A single token from the Syntecnia source.

    Every token carries its source location for full observability —
    when something goes wrong, we can trace it back to the exact character.
    """
    type: TokenType
    value: Any
    location: SourceLocation
    raw: str  # original source text that produced this token

    def __str__(self) -> str:
        if self.type in (TokenType.NEWLINE, TokenType.INDENT, TokenType.DEDENT, TokenType.EOF):
            return f"Token({self.type.name}, {self.location})"
        return f"Token({self.type.name}, {self.value!r}, {self.location})"

File: rust/crates/syntecnia-core/src/tokens.rs
//! Definiciones de tokens de Syntecnia.
//!
//! Port fiel de `syntecnia/core/tokens.py`. Los tokens son las unidades atómicas
//! del lenguaje y reflejan su filosofía: legible, plano, basado en intención.

use std::fmt;

/// Modelo numérico: re-exportado desde [`crate::number`].
/// `Int(i64)` fast-path · `Big(BigInt)` por promoción al desbordar i64 · `Float(f64)`.
/// Replica los enteros de precisión arbitraria de Python.
pub use crate::number::Number;

/// Tipos de token. Espeja `TokenType` del oráculo Python (mismos miembros).
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
pub enum TokenType {
    // -- Literales --
    Number,
    Text,
    BoolTrue,
    BoolFalse,
    Nothing,
    Identifier,

    // -- Flujo --
    When,
    Otherwise,
    Each,
    While,
    Match,
    Is,
    Then,
    AndThen,
    Stop,
    Repeat,
    In,
    With,

    // -- Definiciones --
    Task,
    Give,
    Let,
    Be,
    Set,
    To,
    As,
    Of,
    Type,

    // -- Agentes --
    Agent,
    Spawn,
    Share,
    Observe,
    State,
    Signal,
    WaitFor,

    // -- Capacidades y seguridad --
    Require,
    Allow,
    Deny,
    Sandbox,
    Verify,

    // -- Interacción humana --
    Approve,
    Show,
    Ask,
    Confirm,

    // -- Razonamiento (LLM) --
    Reason,
    Intent,
    Invariant,
    Decide,
    Analyze,
    Generate,

    // -- Manejo de errores --
    Try,
    Recover,

    // -- Soft keywords de servidor HTTP (NO están en KEYWORDS) --
    Serve,
    On,
    Route,
    Auth,
    Requires,
    Expect,
    MaxBody,
    Stream,
    Send,
    MaxStreams,
    RateLimit,
    Per,

    // -- Observabilidad --
    Trace,
    Log,
    Measure,
    Checkpoint,

    // -- Operadores --
    Plus,
    Minus,
    Star,
    Slash,
    Percent,
    Power,
    Equal,
    NotEqual,
    Less,
    Greater,
    LessEqual,
    GreaterEqual,
    And,
    Or,
    Not,
    Assign,
    Arrow,
    FatArrow,
    Pipe,

    // -- Delimitadores --
    LParen,
    RParen,
    LBracket,
    RBracket,
    LBrace,
    RBrace,
    Comma,
    Dot,
    Colon,
    Newline,
    Indent,
    Dedent,

    // -- Especiales --
    Comment,
    Eof,
    Error,
}

impl TokenType {
    /// Nombre del miembro tal como lo expone Python (`TokenType.name`, en MAYÚS).
    /// Los mensajes de error del parser referencian estos nombres, así que son
    /// parte del contrato de paridad.
    pub fn name(&self) -> &'static str {
        use TokenType::*;
        match self {
            Number => "NUMBER",
            Text => "TEXT",
            BoolTrue => "BOOL_TRUE",
            BoolFalse => "BOOL_FALSE",
            Nothing => "NOTHING",
            Identifier => "IDENTIFIER",
            When => "WHEN",
            Otherwise => "OTHERWISE",
            Each => "EACH",
            While => "WHILE",
            Match => "MATCH",
            Is => "IS",
            Then => "THEN",
            AndThen => "AND_THEN",
            Stop => "STOP",
            Repeat => "REPEAT",
            In => "IN",
            With => "WITH",
            Task => "TASK",
            Give => "GIVE",
            Let => "LET",
            Be => "BE",
            Set => "SET",
            To => "TO",
            As => "AS",
            Of => "OF",
            Type => "TYPE",
            Agent => "AGENT",
            Spawn => "SPAWN",
            Share => "SHARE",
            Observe => "OBSERVE",
            State => "STATE",
            Signal => "SIGNAL",
            WaitFor => "WAIT_FOR",
            Require => "REQUIRE",
            Allow => "ALLOW",
            Deny => "DENY",
            Sandbox => "SANDBOX",
            Verify => "VERIFY",
            Approve => "APPROVE",
            Show => "SHOW",
            Ask => "ASK",
            Confirm => "CONFIRM",
            Reason => "REASON",
            Intent => "INTENT",
            Invariant => "INVARIANT",
            Decide => "DECIDE",
            Analyze => "ANALYZE",
            Generate => "GENERATE",
            Try => "TRY",
            Recover => "RECOVER",
            Serve => "SERVE",
            On => "ON",
            Route => "ROUTE",
            Auth => "AUTH",
            Requires => "REQUIRES",
            Expect => "EXPECT",
            MaxBody => "MAX_BODY",
            Stream => "STREAM",
            Send => "SEND",
            MaxStreams => "MAX_STREAMS",
            RateLimit => "RATE_LIMIT",
            Per => "PER",
            Trace => "TRACE",
            Log => "LOG",
            Measure => "MEASURE",
            Checkpoint => "CHECKPOINT",
            Plus => "PLUS",
            Minus => "MINUS",
            Star => "STAR",
            Slash => "SLASH",
            Percent => "PERCENT",
            Power => "POWER",
            Equal => "EQUAL",
            NotEqual => "NOT_EQUAL",
            Less => "LESS",
            Greater => "GREATER",
            LessEqual => "LESS_EQUAL",
            GreaterEqual => "GREATER_EQUAL",
            And => "AND",
            Or => "OR",
            Not => "NOT",
            Assign => "ASSIGN",
            Arrow => "ARROW",
            FatArrow => "FAT_ARROW",
            Pipe => "PIPE",
            LParen => "LPAREN",
            RParen => "RPAREN",
            LBracket => "LBRACKET",
            RBracket => "RBRACKET",
            LBrace => "LBRACE",
            RBrace => "RBRACE",
            Comma => "COMMA",
            Dot => "DOT",
            Colon => "COLON",
            Newline => "NEWLINE",
            Indent => "INDENT",
            Dedent => "DEDENT",
            Comment => "COMMENT",
            Eof => "EOF",
            Error => "ERROR",
        }
    }
}

/// Busca una palabra reservada (hard keyword). Las soft keywords del servidor HTTP
/// (`serve`, `on`, `route`, `auth`, `requires`, `expect`, etc.) **no** están acá:
/// el lexer las emite como `IDENTIFIER` y el parser las reconoce por contexto.
pub fn keyword_lookup(word: &str) -> Option<TokenType> {
    use TokenType::*;
    let ty = match word {
        // Flujo
        "when" => When,
        "otherwise" => Otherwise,
        "each" => Each,
        "while" => While,
        "match" => Match,
        "is" => Is,
        "then" => Then,
        "and_then" => AndThen,
        "stop" => Stop,
        "repeat" => Repeat,
        "in" => In,
        "with" => With,
        // Definiciones
        "task" => Task,
        "give" => Give,
        "let" => Let,
        "be" => Be,
        "set" => Set,
        "to" => To,
        "as" => As,
        "of" => Of,
        "type" => Type,
        // Agentes
        "agent" => Agent,
        "spawn" => Spawn,
        "share" => Share,
        "observe" => Observe,
        "state" => State,
        "signal" => Signal,
        "wait_for" => WaitFor,
        // Capacidades
        "require" => Require,
        "allow" => Allow,
        "deny" => Deny,
        "sandbox" => Sandbox,
        "verify" => Verify,
        // Humano
        "approve" => Approve,
        "show" => Show,
        "ask" => Ask,
        "confirm" => Confirm,
        // Razonamiento
        "reason" => Reason,
        "intent" => Intent,
        "invariant" => Invariant,
        "decide" => Decide,
        "analyze" => Analyze,
        "generate" => Generate,
        // Observabilidad
        "trace" => Trace,
        "log" => Log,
        "measure" => Measure,
        "checkpoint" => Checkpoint,
        // Manejo de errores
        "try" => Try,
        "recover" => Recover,
        // Literales
        "true" => BoolTrue,
        "false" => BoolFalse,
        "nothing" => Nothing,
        // Operadores lógicos
        "and" => And,
        "or" => Or,
        "not" => Not,
        _ => return None,
    };
    Some(ty)
}

/// Posición exacta en el código fuente, para reporte de errores y trazas.
/// `offset` cuenta code points (igual que el `pos` del lexer Python).
#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub struct SourceLocation {
    pub file: String,
    pub line: usize,
    pub column: usize,
    pub offset: usize,
}

impl fmt::Display for SourceLocation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}:{}:{}", self.file, self.line, self.column)
    }
}

/// Valor adjunto a un token. En Python `Token.value` es `Any`; acá lo modelamos
/// como enum según el tipo de token que lo produjo.
#[derive(Clone, Debug, PartialEq)]
pub enum TokenValue {
    /// Literal numérico (NUMBER).
    Number(Number),
    /// Texto: contenido de TEXT, nombre de IDENTIFIER/keyword, lexema de operador
    /// o delimitador, texto de COMMENT.
    Str(String),
    /// Nivel de indentación para INDENT/DEDENT.
    Int(i64),
    /// Sin valor (NEWLINE, EOF).
    None,
}

/// Un token del fuente Syntecnia. Cada token lleva su ubicación para trazabilidad.
#[derive(Clone, Debug, PartialEq)]
pub struct Token {
    pub ty: TokenType,
    pub value: TokenValue,
    pub location: SourceLocation,
    /// Texto original del fuente que produjo este token.
    pub raw: String,
}

impl fmt::Display for Token {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self.ty {
            TokenType::Newline | TokenType::Indent | TokenType::Dedent | TokenType::Eof => {
                write!(f, "Token({}, {})", self.ty.name(), self.location)
            }
            _ => write!(f, "Token({}, {}, {})", self.ty.name(), self.value_repr(), self.location),
        }
    }
}

impl Token {
    /// Contenido string del valor (IDENTIFIER/keyword/TEXT/operador). `""` si no aplica.
    pub fn as_str(&self) -> &str {
        match &self.value {
            TokenValue::Str(s) => s,
            _ => "",
        }
    }

    /// Valor numérico (NUMBER). `Int(0)` si no aplica.
    pub fn as_number(&self) -> Number {
        match &self.value {
            TokenValue::Number(n) => n.clone(),
            _ => Number::Int(0),
        }
    }

    /// Repr estilo Python de `value` (para Display/debug).
    fn value_repr(&self) -> String {
        match &self.value {
            TokenValue::Number(Number::Int(n)) => n.to_string(),
            TokenValue::Number(Number::Big(b)) => b.to_string(),
            TokenValue::Number(Number::Float(x)) => x.to_string(),
            TokenValue::Str(s) => format!("{:?}", s),
            TokenValue::Int(n) => n.to_string(),
            TokenValue::None => "None".to_string(),
        }
    }
}
